llvm/lib/Transforms/Coroutines/CoroFrame.cpp

```diff
@@ -1614,6 +1621,24 @@ struct AllocaUseVisitor : PtrUseVisitor<AllocaUseVisitor> {
       // suspend point between lifetime markers. This should also cover the
       // case of a single lifetime.start intrinsic in a loop with suspend point.
       if (PI.isEscaped()) {
+        // If there is no explicit lifetime.end, then assume the address can
+        // cross suspension points.
+        if (LifetimeEndBBs.empty())
+          return true;
+
+        // If there is a path from a lifetime.start to a suspend without a
+        // corresponding lifetime.end, then the alloca's lifetime persists
+        // beyond that suspension point and the alloca must go on the frame.
+        for (AnyCoroSuspendInst *SuspendInst : CoroShape.CoroSuspends) {
+          SmallVector<BasicBlock *> LifetimeStartsBB;
+          for (IntrinsicInst *II : LifetimeStarts)
+            LifetimeStartsBB.push_back(II->getParent());
+          if (isPotentiallyReachableFromMany(LifetimeStartsBB,
```

Comment:
This looks more expensive since it didn't cache any thing. It would be better if we can reuse the checker above.

Reply:
>  "Is there a path from BasicBlock A to BasicBlock B without going through BasicBlocks { C, D, E, ...}. The current logic caches whether a path may go through a suspend point - it doesn't answer the reverse question

I don't understand this. The reverse form of it should be: there is not a path may go through a suspend point. Then it looks the same with the previous one.

Reply:
The problem is that we have to detect whether there are other `llvm.coro.end`s between `llvm.lifetime.start` and `llvm.lifetime.end` - see for example:

[Block A: `lifetime.start`] -> [Block B: `lifetime.end`] -> [Block C: `coro.suspend`] -> [Block D: `lifetime.end`].

In this case, if we were analyzing A -> D, then we would see that there is a path from A's `lifetime.start` to D's `lifetime.start`, suggesting that we need to put the object on the frame. However, this analysis is incorrect because B ends the lifetime before `coro.suspend`. A correct analysis would need some way to filter out all intermediate `lifetime.end` blocks

This is even more problematic for the case in the original bug where `lifetime.end` was missing. Therefore, we would need to check paths from `lifetime.start` to `return` and `unreachable`. So for example:

[Block A: `lifetime.start`] -> [Block B: `lifetime.end`] -> [Block C: `coro.suspend`] -> [Block D: `unreachable`].

we would run into the same problem because we see that `lifetime.start` -> `unreachable` requires a pass through `coro.suspend`, suggesting that we need to put the object on the frame when we actually don't due to the `lifetime.end`.

Reply:
Got it. Then in another perspective, if `Checker` checks whether there is a path from block A to blockB touching suspending blocks, can we create a similar checker (or refactor the existing one) to check if there is a path from blockA (the block of the a lifeitme.start) to blockB (any suspending blocks) touching ending blocks (the blocks containing corresponding lifetime.end)? Then if there is such a path, we need to put the alloca to the frame. Otherwise we can put it on the stack. It looks like it can solve the two problems you raised.

Reply:
Got it.  I think it will be better to extend `isPotentiallyReachableFromMany` to `isManyPotentiallyReachableFromMany`. e.g., we don't need to iterate the suspending block. The implementation of `isPotentiallyReachableFromMany` looks not complex and should be doable. We can add it in CoroFrame locally or contribute that to CFG.h. 

For cache, yeah, it looks like we can't have an easy one : (

Reply:
This should be done - PTAL